cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/common/ArtifactCacheManager.java

```diff
@@ -0,0 +1,166 @@
+  private static final String CONTENT_TYPE_JSON = "application/json";
+
+  public static void recordCacheUsageForArtifacts(Storage client, String bucket, Set<String> files,
+                                                  String runRootPath, String cachePath) {
+    try {
+      String cachedArtifactsListFilePath = getPath(runRootPath, CDAP_CACHED_ARTIFACTS_LIST_FOR_RUNID);
+      storeCachedArtifactsForRun(client, bucket, files, cachedArtifactsListFilePath);
+      String cacheCountFilePath = getPath(cachePath, CDAP_CACHED_ARTIFACTS_USAGE_COUNT);
+      changeUsageCountForArtifacts(client, bucket, files, cachePath, cacheCountFilePath, 1);
+    } catch (Exception e) {
+      LOG.error("Error in recordCacheUsageForArtifacts", e);
+    }
+  }
+
+  public static void releaseCacheUsageForArtifacts(Storage client, String bucket, String runRootPath,
+                                                   String cachedArtifactsPath) {
+    try {
+      String cachedArtifactsListFilePath = getPath(runRootPath, CDAP_CACHED_ARTIFACTS_LIST_FOR_RUNID);
+      Set<String> files = getCachedArtifactsForRun(client, bucket, cachedArtifactsListFilePath);
+      String cacheCountFilePath = getPath(cachedArtifactsPath, CDAP_CACHED_ARTIFACTS_USAGE_COUNT);
+      changeUsageCountForArtifacts(client, bucket, files, cachedArtifactsPath, cacheCountFilePath, -1);
+    } catch (Exception e) {
+      LOG.error("Error in releaseCacheUsageForArtifacts", e);
+    }
+  }
+
+  private static void changeUsageCountForArtifacts(Storage client, String bucket, Set<String> files, String cachePath,
+                                                   String filePath, int changeValue) {
+    if (files.isEmpty()) {
+      return;
+    }
+    BlobId blobId = BlobId.of(bucket, filePath);
+    int maxRetries = 10;
+    for (int i = 0; i < maxRetries; i++) {
+      try {
+        Blob blob = client.get(blobId);
+        Map<String, Integer> artifactCount;
+        if (blob == null || !blob.exists()) {
+          LOG.debug("ArtifactsCacheCounter File doesn't exist! Creating..");
+          artifactCount = new HashMap<>();
+          BlobInfo createBlob = BlobInfo.newBuilder(blobId).setContentType(CONTENT_TYPE_JSON).build();
+          try {
+            client.create(createBlob, GSON.toJson(artifactCount).getBytes(StandardCharsets.UTF_8),
+                          Storage.BlobTargetOption.doesNotExist());
+          } catch (StorageException e) {
+            if (e.getCode() != HttpURLConnection.HTTP_PRECON_FAILED) {
+              throw e;
+            }
+          }
+          blob = client.get(blobId);
+        }
+        artifactCount = GSON.fromJson(new String(blob.getContent(), StandardCharsets.UTF_8),
+                                      new TypeToken<Map<String, Integer>>() {
+                                      }.getType());
+        for (String file : files) {
+          int count = artifactCount.getOrDefault(file, 0) + changeValue;
+          if (count <= 0) {
```

Comment:
It should not go negative. We probably need to log error if it happens.

Reply:
Added a log here.